### PR TITLE
fix: fetch revision only then fallback to default refspec

### LINKED ISSUE #4935: Allow referring to non-default remote references
# Summary

In the application revision field, support specifying any valid remote ref

# Motivation

I'd like to be able to point at a GitHub pull request by specifying a revision of e.g. `refs/pull/168/head`

# Attempted Workaround

I tried to add `remote.origin.fetch=+refs/pull/*/head:refs/remotes/origin/refs/pull/*/head` and using a revision of `origin/refs/pull/168/head` to the repo-server git config, but that didn't seem to work. I guess ArgoCD uses some git implementation that ignores the git config?

### PR DESCRIPTION
**Pull request by yujunz**
Fixes #4935

### CODE REVIEW

**Comment by yujunz** [comment]
Error messages

```
 time="2021-02-25T08:10:01Z" level=info msg=Trace args="[git fetch origin --tags --force]" dir="/tmp/argocd@****" operation_name="exec git" time_ms=9547.973753
 time="2021-02-25T08:10:01Z" level=info msg="git checkout --force 45a223fedbb93eeae70820d534257f90a85f961c" dir="/tmp/argocd@****" execID=QfaG4
 time="2021-02-25T08:10:01Z" level=error msg="`git checkout --force 45a223fedbb93eeae70820d534257f90a85f961c` failed exit status 128: fatal: reference is not a tree: 45a223fedbb93e
```

**Comment by yujunz** [comment]
Any comments? @alexmt

**Comment by alexmt** [comment]
Sorry, PR slip off radar.

**Comment by snuggie12** [comment]
@alexmt apologies if a faux pas, but do you have an ETA on this one? Just wanted to make sure this didn't fall through the cracks again due to the holidays.

**Comment by yujunz** [comment]
Flake? The failure seems not related to the change.

```
2022-01-24T03:00:04.1594119Z time="2022-01-24T03:00:04Z" level=debug msg="Application 'test-hook-before-hook-creation-failure' operation terminating\n" duration=204.602806ms execID=47273
2022-01-24T03:00:04.1595605Z time="2022-01-24T03:00:04Z" level=info msg="expectation succeeded: no error and output contained ''"
2022-01-24T03:00:04.1659029Z time="2022-01-24T03:00:04Z" level=info msg="pending: operation phase should be Failed, is Running"
2022-01-24T03:00:07.1712699Z time="2022-01-24T03:00:07Z" level=info msg="pending: operation phase should be Failed, is Succeeded"
2022-01-24T03:00:10.1777595Z time="2022-01-24T03:00:10Z" level=info msg="pending: operation phase should be Failed, is Succeeded"
2022-01-24T03:00:13.1833939Z time="2022-01-24T03:00:13Z" level=info msg="pending: operation phase should be Failed, is Succeeded"
2022-01-24T03:00:16.1909338Z time="2022-01-24T03:00:16Z" level=info msg="pending: operation phase should be Failed, is Succeeded"
2022-01-24T03:00:19.1918959Z ##[error]    hook_test.go:351: timeout waiting for: operation phase should be Failed, is Succeeded
2022-01-24T03:00:19.1920454Z --- FAIL: TestHookBeforeHookCreationFailure (18.96s)
2022-01-24T03:00:19.1920913Z FAIL
2022-01-24T03:00:19.1983630Z FAIL	github.com/argoproj/argo-cd/v2/test/e2e	651.703s
2022-01-24T03:00:19.2048239Z FAIL
```

**Comment by yujunz** [comment]
Failed on the same case

```
Error:     hook_test.go:351: timeout waiting for: operation phase should be Failed, is Succeeded
--- FAIL: TestHookBeforeHookCreationFailure (18.93s)
FAIL
FAIL	github.com/argoproj/argo-cd/v2/test/e2e	643.169s
FAIL
```

Is the `master` branch broken?

File: reposerver/repository/repository.go [hunk 1]
```diff
@@ -1574,25 +1574,25 @@ func checkoutRevision(gitClient git.Client, revision string) error {
...
-		}
-		err = gitClient.Checkout("FETCH_HEAD")
-		if err != nil {
-			return status.Errorf(codes.Internal, "Failed to checkout FETCH_HEAD: %v", err)
-		}
-	} else {
+	err = gitClient.Fetch(revision)
+
+	if err != nil {
+		log.Warnf("Failed to fetch revision %s: %v", revision, err)
```
Comment (alexmt) [inline reposerver/repository/repository.go:1580]: Please change it to `log.Info`. We are trying to use 'warning' logs only for something that the operator should act on.

File: util/git/client.go [hunk 2]
```diff
@@ -335,7 +335,7 @@ func (m *nativeGitClient) Fetch(revision string) error {
 
 	var err error
 	if revision != "" {
-		err = m.runCredentialedCmd("git", "fetch", "origin", revision, "--tags", "--force")
+		err = m.runCredentialedCmd("git", "fetch", "origin", revision)
```
Comment (alexmt) [inline util/git/client.go:338]: Can you please clarify why this change is necessary? Can we rollback this line please unless it is really needed?

**Review by alexmt — CHANGES_REQUESTED** [review]

File: reposerver/repository/repository.go (hunk 1 above)
Comment (yujunz) [reply reposerver/repository/repository.go:1580]: Done

File: util/git/client.go (hunk 2 above)
Comment (yujunz) [reply util/git/client.go:338]: No need to fetch tags on explicit revision. This reduce a bit the pack size on large repository with many tags.

Feel free to revert the change though.

Comment (alexmt) [reply util/git/client.go]: Sounds good. Reverted this change. The rest looks good! Merging once CI is green.

**Review by alexmt — APPROVED** [review]

**Comment by crenshaw-dev** [comment]
> If performance is a concern, we can add an explicit setting for each repository.

@yujunz can you expand on this? I'm observing performance issues (lots of large packfiles) for some repos. But I don't know enough about git to understand why this happens.

**Comment by yujunz** [comment]
Not sure what kind of pack files from your repository. This patch mainly reduces the total refs to fetch. It is not helping when the repository has many large files or long commit history.

**Pull request merged**